### ISSUE TITLE
cargo publish should be opt-in

### ISSUE THREAD

**Issue by jmaargh** [rust-lang/cargo#6153]
Publishing, whether to crates.io or elsewhere, should be opt-in rather by on-by-default. Although rare, people [do sometimes publish accidentally](https://stackoverflow.com/questions/52639162/how-to-delete-a-published-crate-from-crates-io). Given that the default is currently to publish to crates.io publicly, this is potentially dangerous for using cargo in a corporate closed-source environment. My guess is that even most always-intended-to-be-open-source projects would rather keep things private until they're ready for some sort of release.

I see several options for making this work:

1. Have `cargo new` add `publish = false` or `publish = []` or equivalent in the default `Cargo.toml`. This is probably the most minor change and I can't imagine this will break any existing uses.

2. Change the behaviour so that `publish = true` in `Cargo.toml` is _required_ for `cargo publish` to work, with an error message explaining this. This might break workflow for some existing projects, but it would be a very easy fix.

3. Change the behaviour of `cargo publish` so that if there is no `publish = True` in `Cargo.toml` it interactively asks whether you're sure. Possibly with a `-y`/`--yes` option (a la `apt install`) to automatically say "yes".

Personally, I'd prefer both 1 and 2. I'd be happy to make a PR if people like this idea.

Originally raised as an idea on #6123, but was decided it belonged as a separate issue.

**Comment by dwijnand** [comment]
Is this a distinct issue to that being discussed in #6123?

**Comment by yonni12** [comment]
Yes, I posted a comment there and people asked me to create as separate issue. Sorry, should have linked.

**Comment by withoutboats** [comment]
My initial inclination is against doing something like this, but I could be convinced.

There are a few barriers already against accidentally publishing:

1. The `cargo publish` command has no other purpose, and none of our other major commands are similar, making users less likely to accidentally run `cargo publish` when they meant to run something else.
2. The publication will fail if you aren't logged in, making it less likely for you to successfully publish something when you are new to Rust & haven't yet learned what `cargo publish` is for. That is, the first time you publish you have to actively try to by creating a crates.io account, so we avoid the problem of people running commands without understanding them.
3. The crate created by `cargo new` will be rejected by crates.io because it lacks a license and description field. You have to have set both of those up before you can publish.
4. `cargo publish` fails if your crate doesn't build or if your VCS repository is dirty.
5. If you're publishing a subsequent version of a crate, `cargo publish` will fail if you haven't updated the version number to the new version.

My own experience of publishing is that I almost always run afoul of one of the existing checks when I go to publish - usually I haven't set a license or description field.

Often the code you don't want to publish is a project you intend to publish someday (and you may have already published a previous version of it). Its perfectly possible to set the publish bit and then run `cargo publish` by mistake later. That is, this only prevents an accidental publish of a completely new crate, and only once all of the other roadblocks mentioned earlier have been hurdled by the user.

**Comment by yonni12** [comment]
I agree that accidentally publishing is probably a nice issue, but I don't see how yours are arguments _against_ this proposal, so much as they are arguments that in your experience it's _unnecessary_.

One way to think about it is this: if we're going to have a `publish` flag, then it should default to the "safe" state of `false` (you never lose anything -- except a couple of seconds -- by not publishing when you meant to). If the `publish` flag is so unnecessary, then it should be removed altogether. Right now, there is a flag whose only purpose is to save you from yourself that defaults to the "unsafe" option.

As for the other barriers, I think they should all remain, but none of them are as explicit as "I don't want this published". For example, you might choose to not set a license, or not log in, simply because you don't want to publish. But in a year's time, you're unlikely to remember that. The advantage of the `publish` flag is that the intentionality is explicit.

**Comment by joshtriplett** [comment]
My initial feeling is -0 (weakly opposed). I don't feel *strongly* about this, it just feels like a bit of an unnecessary stumbling block in a new project. I can absolutely understand the desire to make sure people don't publish accidentally, however.

**Comment by sfackler** [comment]
@ralfbiedert that's the current behavior via the `[package] publish = false` setting.

**Comment by LucianoBestia** [comment]
There are actions that are reversible and actions that are irreversible.  
`cargo publish`  
is terribly easy to write, run and misunderstand.  
Later is not possible to delete a published crate.  That is very irreversible !
That is disproportional !
A simple easy security measure should be enought:  require the name of the crate to publish:  
`cargo publish --crate my_crate_name`  
That is more complicated to write but is proportional to the gravity of the action.  

Another idea is to allow 5 minutes staging time to delete the published crate on crates.io.  
In that time it should not be visible to others.  
And maybe a command `cargo publish --force --crate my_crate_name'  
for the people that need it published pronto.

**Comment by LucianoBestia** [comment]
Publishing a new crate or a new stable version of a crate is pretty rare.  
And it needs a lot of preparation and thinking through.  
An unintentional publish can have very grave consequences.  
Because of that the command could be a little more engaging for the developer.  
For example requiring this arguments:
`cargo publish --crate my_crate_name --vers 1.0.0. --reg crate-io`
This protects the developer to not publish unintentionally the wrong crate or the wrong version to the wrong registry.  
Having dangerous default values can lead to secrets or private code been published publicly.  
In a split second, there are robots that will make a copy of that code.  
And in a few minutes, they will exploit a password or a secret token.  
**These are very bad consequences.**  
We can easily prevent that from happening.

**Comment by jmaargh** [comment]
@LucianBuzzo The point that people might accidentally (or lazily) have secrets in code they don't intend to publish is a good point I hadn't thought of.

Personally, I find the current behaviour at odds with the usual approach to safety in Rust. Publishing code (especially to crates.io, which has no easy removal mechanism) is an "unsafe" action, therefore it should require some explicit opt-in on the part of a human.

**Comment by mcobzarenco** [comment]
> Another idea is to allow 5 minutes staging time to delete the published crate on crates.io.

That's almost implemented via rustc compilation times -- there's usually plenty of time after typing `cargo publish` to change your mind :-)

>  cargo publish --crate my_crate_name
The only time I ran `cargo publish` by mistake is pressing arrow up and rerunning the command from shell history -- this would not guard against that.

I tend to (weakly) agree with the opinions already expressed that the current behaviour is not too bad -- there are many "irreversible" commands one can run (e.g.`rm -rf`) -- given that `cargo publish` doesn't have any other uses, the status quo seem appropriate. 

From the 3 solutions proposed by the OP, 1 seems particularly unappealing as the default should be open source, not a corporate environment (writing this as someone who mostly uses Rust in a closed source corporate env).

**Comment by LucianoBestia** [comment]
Sorry, you didn't mention my last proposal - all args mandatory:
'cargo publish --crate my_crate_name --vers 1.0.0. --reg crate-io'
That solves all current safety/security/clumsiness issues:
1. Cannot type 'cargo publish' by mistake and expect a guide/help for the utility. Instead it goes for the kill.
2. Cannot inadvertently call 'cargo publish' in the wrong project.
3. Cannot mistakenly publish the wong version.
4. Cannot forget to explicitly choose the  crate database
I suppose developers use the bash history for all the frequent commands. Also this.

**Comment by LucianoBestia** [comment]
There are 2 extreme philosofies and a big gray area between:
1. Most security and less comfort
2. Most comfort and less security.

Javascript and dynamic languages are very comfortable. You dont need to think about types or lifetimes. But that is also less secure when later refactoring. The compiler will not help you to spot errors.
I choose the Rust programming language exactly because it gives me all the errors, warnings and alerts in compile time. I choose to work uncomfortably to achieve more security.
So I expect that philosofical standoint also for other parts of the Rust ecosystem.
Here explicitly debating about `cargo publish`.

**Comment by NTmatter** [comment]
As an alternative to changing defaults, would it be more palatable to allow users to specify a behavior for new projects in their [config.toml hierarchy](https://doc.rust-lang.org/cargo/reference/config.html#hierarchical-structure)? This is already an option for  name, email, and vcs.
```toml
[cargo-new]
# Avoid publishing my hello_world projects
publish = false
email = "test@example.com"
```

Would it make more sense to allow specifying _all_ defaults for cargo-new's manifest, bypassing the special-casing for name/email/vcs?
```toml
[cargo-new.default-manifest]
[cargo-new.default-manifest.package]
edition = 2018
publish = false
license = "MIT"

[cargo-new.default-manifest.badges]
maintenance = { status = "experimental" }
```

**Comment by SirWindfield** [comment]
Maybe an approach like `standard-version` from the Javascript ecosystem is a nice fit here. By default, running the publishing process is actually a dry-run. You have to set the `-w,--write` CLI flag to make actually do the publishing.

**Comment by 0x6273** [comment]
How about an interactive prompt when publishing a new crate? It would show the crate name and what registry index it will be published to, and it could prompt for the crate name instead of asking Yes/No, to stop the people that won't check the crate name from shooting themselves in the foot.

The prompt could be disabled with a cli opt, in config.toml, and maybe be disabled by default if there's no tty.

**Comment by jmaargh** [comment]
So, this conversation fizzled out. Having reviewed the above discussion, my argument is that:

Priors:
a) Publishing (to wherever, but especially to crates.io) is "unsafe" and almost certainly irreversible
b) Publishing accidentally is rare
c) Workflow stumbling blocks are to be avoided where possible
d) Publishing code containing secrets is potentially catastrophic
e) Publishing to the wrong repo is potentially catastrophic

Thus, we have a risk of something very rare but potentially also very damaging (always difficult to deal with proportionally) and also a risk that "fixing" it may introduce annoyances. My suggestion is therefore to de-risk by making the unsafe option explicitly opt-in in the simplest way possible, but also to use the error message(s) to provide useful feedback about how to publish (thus turning the stumbling block instead into a teaching moment).

Suggestion:
a) Default the `publish` flag internally to `false` (i.e., absence of the flag is interpreted as "false", keep the flag absent in the default configuration)
  - Any existing published project is already likely to have explicitly set `publish = true`, for those that haven't it's a very simple one-line change
  - Nearly every instance of accidental publishing will be caught by this, because somebody has to have explicitly opted-in at some point
b) When `cargo publish` fails, show a loud warning in stdout that teaches the user what's gone wrong and what they were about to do
  - Something like `Warning: failed attempting to publish to [insert repository here, highlighted]`
  - Followed by a checklist of required steps (like `[x] logged in with cargo login`, `[ ] no licence specified in Cargo.toml`, `[ ] publish flag not set in Cargo.toml`)
  - A link at the end to docs on how to properly publish a crate (e.g. https://doc.rust-lang.org/cargo/commands/cargo-publish.html)

I hope this strikes a good balance between safety and ergonomics (doesn't that sound like Rust :wink: )

**Comment by vadixidav** [comment]
I have something to add here that I haven't seen discussed which has almost slipped me up several times.

The [cargo release](https://github.com/sunng87/cargo-release) project, which is used by quite a few people, will publish your crate when cutting a new version if `publish = false` is not set. I use this as part of private repositories that should not be published, but I nonetheless need to maintain mono repo versioning, and adding a new project in that mono repo can inadvertently lead to publishing it accidentally. So far, I have had to be judicious to always add the `--dry-run` flag, but I have come close on several occasions to publishing a new crate, or at least I feel that I would have had I not looked at the dry run. This has concerned me enough that I did come looking to raise an issue precisely like this one.

**Comment by JanBeh** [comment]
Disclaimer: I'm new to publishing crates (I uploaded my first crate yesterday). I am also worried about accidentally publishing something that I don't want to publish (and apparently `cargo publish` even uploaded an (untracked) backup file that happend to be in my local repository). I currently delete my API key on crates.io (to be sure) or have to remove my local credentials to avoid accidental publishing (e.g. due to execution of a wrong command in the shell history). I would definitely like cargo to be more conservative when publishing something permanently. Adding `[package] publish = false` to the `~/.cargo/config.toml` doesn't seem to be respected either. Ideally, I would expect such a command to **always** interactively ask me. I personally don't see a need to be able to automate publication (but maybe there are cases where it's necessary). Either way, it shouldn't be the default to not even ask.

**Comment by Goffen** [comment]
I almost accidentally pushed my crate for my employer to crates.io because I've misunderstood something with the configuration of registry:s

First time publishing a crate should maybe prompt some extra security. Like "Are you sure these files ..... will be uploaded to crates.io/xx/yy" ?!

Thank god I've not configured an access token to crates.io

**Comment by epage** [comment]
Shouldn't have marked this closed by #12786.  We've now made `package.publish` opt-in if you omit `package.version` but most projects won't omit it unless the person knows that and explicitly does so.  Whether that is sufficient for closing this, I'm unsure.

**Comment by jmaargh** [comment]
Thanks for re-opening. I agree that #12786 (while a positive change) does not close this issue. Not least because "this is now opt-in if you do this non-default thing" is the definition of opt-out, not opt-in.

Previous commenters have pointed to other barriers to publishing (logging in, licence set, description set, etc.), but I think my main problem with all of them is that they are proxies and I don't think that Cargo should lean on implicit proxies for interaction design that can stop a footgun. I think it should explicitly design against footguns, which defaulting `package.publish` to `false` when omitted would do. Users would have to engage in a specific positive action of setting this field.

Again, happy to write a PR for this if the maintainers can be persuaded that it is the right thing to do.

**Comment by bestia-dev** [comment]
We already established that the command `cargo publish` can be catastrophic in some special circumstances (containing secrets, wrong repo, wrong version, no possibility to delete it after publish).
Let's make a comparison with the `rm` command. If we type only `rm` it does not remove all the files from the system. It returns the error `rm: missing operand`. That is pretty secure. If you know what you are doing you know you have to explicitly say what files to remove. Same here. If you really want to publish, you know how to add the correct registry, project name and version number to the command. These arguments should be mandatory.  
Publishing must be carefully considered. It is not a simple impulse.
That would be much better then the `publish=false` opt-out that we have today.

For ***backward compatibility*** an explicit `publish=true` and `repository=crates.io` in Cargo.toml could mean you don't need additional arguments for the command `cargo publish` because they are already defined in Cargo.toml.  
I think that the list "include" should also be mandatory for publishing. Usually you don't want to publish examples, tests and sample data:

```toml
# publish to registry crates.io. Only this files.
publish = true
registry= crates.io
include = [
    "Cargo.toml",
    "LICENSE",
    "README.md",
    "src/*"
]
```

**Comment by tesaguri** [comment]
> I think that the list "include" should also be mandatory for publishing. Usually you don't want to publish examples, tests and sample data:

Aren't `examples` and `tests` technically part of a package? For example, the following works:

```command
$ cargo add bitflags && cargo check -p bitflags --example fmt
```

I don't know whether omitting those targets would do any harm in practice, but at least that doesn't seem to be no-op.

Anyway, I think that making `publish` opt-in should suffice for preventing irreversible mistakes and that mandating `include` field would cause too much churn to the ecosystem while its benefit is negligible for most projects.

---

I think the option 2. proposed in OP is the most uncontroversial change for starter, as there is a precedent for changing the default manifest, that is, `cargo new --bin` has become the default instead of `cargo new --lib` in toolchain [v1.25](https://blog.rust-lang.org/2018/03/29/Rust-1.25.html).

I wonder if there is a motivation *against* this approach: in particular, does <https://github.com/rust-lang/cargo/issues/6153#issuecomment-427902074> still stand?

**Comment by jmaargh** [comment]
After some discussion [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/cargo.20publish.20should.20be.20opt-in), here's what I propose:

1. This issue (#6153) simply narrows in scope to adding `publish = false` to the default `Cargo.toml` generated by `cargo new`, together with a short comment linking to documentation about publishing. This is a simple change which is a big step in the right direction, doesn't require an edition boundary as it's not a breaking change, and is unlikely to break many (if any) workflows.
2. A new issue is created for implementing a breaking change to default behaviour on the next edition boundary. Details drawn from the current discussion and I'll leave the details to the new issue (though I do have a specific permutation of the proposals in mind).

Per the [contributing](https://github.com/jmaargh/cargo/blob/master/CONTRIBUTING.md) guide, if this issue is marked as `accepted`, I'll make a PR for (1) and a new issue for (2).

**Comment by ReveStobinson** [comment]
I'm not a Rust developer at all, but have recently found myself responsible for building some CI jobs around a Rust package with python bindings, and have encountered a similar issue to the one stated in this feature request. I just wanted to add my 2 cents about how this could be handled, based on how we do things over in Python land.

[PyPI](pypi.org) is our package index (equivalent to crates.io), and it is (primarily) accessed via `pip` (our `cargo` equivalent, at least for this example). In our case, the accidental publication rejection is actually **built into PyPI**, rather than being part of the `cargo` tool itself. 

PyPI has the notion of [classifiers](https://pypi.org/classifiers/), which can be specified in your package configuration file, and are helpful for searching packages based on topic, OS, License, intended audience, etc. But it also includes a special classifier: `"Private :: Do Not Upload"`, which will cause PyPI to reject your package upload request even if it successfully authenticates. (It will actually reject anything with a `"Private ::"` classifier)

Might it be feasible to instead add a similar check to crates.io, based on the `[package.categories]` field of the `Cargo.toml` file? It would require adding a `"Private :: Do Not Upload"` or similar category to the acceptable list, and adding logic to check that first before any code is actually stored in the registry. The upshot of an approach like this is that it could still allow for the package to be uploaded to private and locally-hosted crate registries like [kellnr.io](kellnr.io) or JFrog Artifactory, since they wouldn't be performing that check.

It also prevents the need for setting defaults that might not be intuitive, or could be confusing or inconvenient for users trying to publish a new package quickly.

**Comment by epage** [comment]
Unsure how much of a benefit we get from switching to server side validation from client side validation especially since there is one canonical publishing tool./.  People can still publish on their own but they are taking on risks.  Such a system would also be redundant with the existing client-side metadata which supports (1) no publish, (2) publish to any where, and (3) publish to specific registries

**Comment by epage** [comment]
Many things have changed, in both directions, since this was opened.

First off, some smaller ways of getting this more explicitly
- Leaving off the version field makes a package unpublishable.
- `workspace.package.publish = false` will be auto-inherited by `cargo new`

The motivation given is dealing with accidental publishes.  One rebuttal is the steps necessary to publish:
- user explicitly chose to publish by running the command
  - Its easy to accidentally be in the wrong directory and `cargo publish` is a irreversible action
  - With 1.90, we'll have workspace publishing and it would be easy to accidentally include a new local-only package since the last publish
- being logged in:
  - but if you are logged in for one package, you are logged in for all
- `cargo new` creates an unpublishable package due to lack of metadata
  - if that metadata is in `workspace.package`, then `cargo new` will inherit it making the package publishable
- doens't build or dirty vcs
  - at least for myself, both of those will usually succeed.  Even more so for the `jj` folks.
- will fail for subsequent publishes
  - I guess if I make the mistake once, I'll get an error rather than being able to do it again?  But then that means I did make the irreversible action
  - but wrapper tools like `cargo release` will bump the version so it won't fail

For the "accidental publish" concern, I would personally solve that by making dry-run the default.  However, I don't see a path for us changing a CLI default like that.

Another change is that we have tools using `cargo publish` as plumbing, providing more features.  For example, `cargo release` which makes dry-run the default.

Cases have come up within Cargo where it would be helpful to know definitively that a package will be published rather than the current state where you don't know.
- If we get a lint for #4840 that runs during `cargo build` (ideal to catch problems when resolving is cheapest / easiest), we would not to know if the package truly will be published which we don't know with the default, see also https://github.com/rust-lang/cargo/issues/4377#issuecomment-3299277127
- Published packages require that the package operate in isolation.  There are workflows (e.g. shared assets) that involve not working in isolation and not publishing is fine.  The only way to know if a package is not published is the user explicitly saying so despite their intent may be to never publish it.  See also https://github.com/rust-lang/cargo/issues/3946#issuecomment-3224923131

For myself, I think it would be useful to know definitively that a publish is intended rather than our quasi-state today with the default.

One potential route for this:
- have `cargo new` inject `package.publish = false` as a short term fix
- change the default for `package.publish` on an edition

**Path forward**
- A survey existing tools was requested at [#t-cargo > cargo publish should be opt-in @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/246057-t-cargo/topic/cargo.20publish.20should.20be.20opt-in/near/402022096)
- Due to the ecosystem impact, this likely should be an RFC

**Comment by epage** [comment]
If publish becomes opt-in, then we can give a parse error if a non-dev dep lacks a version when `publish = true`, see also #16229.
